pkg/backend/filestate/backend.go

```diff
@@ -233,19 +237,102 @@ func New(ctx context.Context, d diag.Sink, originalURL string, project *workspac
 	}
 	backend.currentProject.Store(project)
 
+	projectMode := true
 	switch v := pulumiState.Version; v {
 	case 0:
 		backend.store = newLegacyReferenceStore(wbucket)
+		projectMode = false
 	case 1:
 		backend.store = newProjectReferenceStore(wbucket, backend.currentProject.Load)
 	default:
 		return nil, fmt.Errorf(
 			"state store unsupported: 'Pulumi.yaml' version (%d) is not supported "+
 				"by this version of the Pulumi CLI", v)
 	}
+
+	if !projectMode {
+		return backend, nil
+	}
+
+	// If we're in project mode warn about any old stack files.
+	refs, err := newLegacyReferenceStore(wbucket).ListReferences()
```

Comment:
Might this be an expensive calculation if they're referencing remote blob storage?

Reply:
Yeah, it could be a remote list operation.
And yeah, I do dislike that it'll always check once you're in project mode.

We could maybe make this a side-effect of the regular list stacks operation: when we're listing the bucket anyway.

Reply:
This check was moved to #12437, and we added an env-var opt-out for it.